### DISCUSSION TITLE
OpenGl in Avalonia. Questions about teapot.bin from GpuInterop.

### DISCUSSION THREAD

**Discussion by PITON369** [AvaloniaUI/Avalonia#16680] [Q&A]
I cloned a repository for learning how to work with OpenGl in Avalonia. I don't understand what extension the 3D model of a teapot.bin has, because .bin is a very common format and I don't know how to relate it to anything. I would like to add a Blender 3D model (in .obj format) to my project that will constantly rotate itself. I tried to repeat the code from the project on my own, but it turned out to be difficult, because there are many interrelations in the form of services, interfaces, patterns, etc. It takes a long time and is difficult to figure out what is connected to what and how. Most likely, I need literally one or two classes of several dozen lines, without everything that is needed for interrelations in the project and scaling. Usually it is easier to find a tutorial with what is needed and repeat it, but I did not find anything suitable. To make it more convenient and specific, here are my questions:
1) Are there any other small projects where I can see how to implement 3D model rotation? Or maybe tutorials/examples?
2) Can anyone please tell me or give me a link to documentation on the .bin format. It's very hard to google because bin is a very common combination of characters and the results are on unrelated topics. All I could find was that it is a binary format. I couldn't open it in a 3D editor. Is it possible to convert models of other formats to bin? Or is it possible to write code for other formats? Maybe there is some converter?
3) Are there any tips on how to best act in situations when I need to take a piece from another project? Now I can either mess around with the project for a long time until I understand all the connections, or I find a tutorial/do it myself. Maybe there are some tips/methods on how to act faster and easier? For example, for some structures, can we make some stubs or some other typical things? Or is there some algorithm for how to act in such a situation?

**Comment by danipen** [comment]
This may help https://github.com/danipen/AvaloniaOpenGLTutorial

**Reply by PITON369** [reply]
I have successfully built the solution, but when I run any project I don't see the figure. Just an empty window. I see UI in 22 tutorial, but I don't see model. I tried running several different tutorials. Maybe I need some libraries or something else? P.S. I have problems with YouTube and I can't watch videos from it.

**Reply by danipen** [reply]
Which platform are you using? It works fine on macOS, which is what I'm targeting.
https://github.com/user-attachments/assets/85c8e004-3134-4158-a87b-9c88a581ba6e


Anyway, I suggest updating to a more recent version of Avalonia. The samples are based on an outdated version (`11.0.0-preview1`).

**Reply by danipen** [reply]
Regarding your question about formats: The samples use [Assimp.net](https://bitbucket.org/Starnick/assimpnet/src) to read [various formats](https://github.com/assimp/assimp/blob/master/doc/Fileformats.md) and convert them into a common one, which is then used to draw the vertices.

**Reply by PITON369** [reply]
I use Windows for development and Linux for publishing. Where you have a helicopter, I have an empty white window. I'll try your recommendations. Thanks!

**Reply by PITON369** [reply]
I tried updating. When updating avalonia there are too many errors. But everything seems to work on these versions too.
The project has a resource loader where the model is loaded. It's code from ResourseLoader.cs:
public static Scene LoadHelicopterModel()
{
    return ImportFromResource("helicopter.fbx");
}
When executing the code in debug, this code is not executed. I think this should not be the case, am I right? 
For example I choose Tutorial22. I expect OpenGlControl to start OnOpenGlInit. Because this is the only place where the resource loader is called in project. But this code is not executed too. And this function have 0 references. I am happy to provide any details you may need to answer these questions.

**Reply by danipen** [reply]
Well, probably some things changes from 1.0.0-preview1 to latest. So probably some parts of the code should be updated accordingly. I'm afraid I don't have enough resources to take a look into it now...
